--- origsrc/gcc-4.5.3/gcc/config/i386/cygwin.h	2011-07-10 00:52:31.709323700 -0500
+++ src/gcc-4.5.3/gcc/config/i386/cygwin.h	2011-07-10 01:24:24.630654600 -0500
@@ -126,7 +126,8 @@ along with GCC; see the file COPYING3.
   %{shared|mdll: --enable-auto-image-base -e \
     %{mno-cygwin:_DllMainCRTStartup@12} \
     %{!mno-cygwin:__cygwin_dll_entry@12}}\
-  %{!mno-cygwin:--dll-search-prefix=cyg -tsaware}"
+  %{!mno-cygwin:--dll-search-prefix=cyg \
+    %{!shared: %{!mdll: --large-address-aware --tsaware}}}"
 
 /* Allocate space for all of the machine-spec-specific stuff.
    Allocate enough space for cygwin -> mingw32  munging plus
